app/services/youtube_service.rb

```diff
@@ -1,36 +1,39 @@
 class YoutubeService
-  attr_reader :keywords
+  attr_reader :keywords, :tags
 
-  def initialize(keywords)
-    @keywords = keywords
+  def initialize(params)
+    @keywords = params[:keywords]
+    @tags = params[:tags]
   end
 
   def videos
-    get_json[:items]
+    videos = get_from_cache_or_youtube(keywords)
+    videos = get_from_cache_or_youtube(tags) if videos[:items].empty?
+    videos
   end
 
   private
 
-  def get_json
-    if $REDIS.get(keywords)
-      json = eval($REDIS.get(keywords))
-    else
-      @response ||= conn.get
-      json = JSON.parse(@response.body, symbolize_names: true)
-      $REDIS.set(keywords, json, ex: 1.day)
-    end
+  def get_from_cache_or_youtube(query)
```

Comment:
This is semantic enough for me to understand in a different language so thank you!